src/python/pants/engine/rules.py

```diff
@@ -26,14 +27,79 @@
 
 
 class _RuleVisitor(ast.NodeVisitor):
-  def __init__(self):
+  """Pull `Get` calls out of an @rule body and validate `yield` statements."""
+
+  def __init__(self, func, frame, parents_table):
     super(_RuleVisitor, self).__init__()
     self.gets = []
+    self._func = func
+    self._frame = frame
+    self._parents_table = parents_table
+    self._yields_in_assignments = set()
+
+  class YieldVisitError(Exception):
+    def __init__(self, node, func, frame, msg, *args, **kwargs):
+      filename, line_number, _, context_lines, _ = inspect.getframeinfo(frame, context=4)
+      err_msg = ("""\
+In function {func_name}: {msg}
+{filename}:{line_number}:
+{context_lines}
+
+The invalid `yield` statement (in the body of the above function) was: {node}
+""".format(func_name=func.__name__, msg=msg,
+           filename=filename, line_number=line_number,
+           context_lines=''.join(context_lines).strip(),
+           node=ast.dump(node)))
+      super(_RuleVisitor.YieldVisitError, self).__init__(err_msg, *args, **kwargs)
+
+  def _maybe_end_of_stmt_list(self, attr_value):
```

Comment:
pydocs on both `_maybe_end_of_stmt_list` and `_stmt_is_at_end_of_parent_list` that explain the condition they're detecting (or refer to a class pydoc that explains it) would be good. The exception error message is very good, but a lower level explanation of the implementation would be great.

Reply:
Will do!